data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/criteria/QueryCriteriaMethodMatch.java

```diff
@@ -255,7 +255,7 @@ protected MethodMatchInfo build(MethodMatchContext matchContext) {
                 }
                 countQuery.join(joinPath.getPath(), joinType, null);
             }
-            countQueryResult = queryBuilder.buildQuery(countQuery);
+            countQueryResult = queryBuilder.buildQuery(annotationMetadataHierarchy, countQuery);
```

Comment:
This is the cause of the issue, when building count query method annotation metadata is not passed so we don't have information about `@Where` annotation. 
This was probably not done by mistake, maybe wanted to avoid `@Join` or something else but I tested and count query with this change still does not consider joins so should be safe. Anything else @dstepanov could perhaps be the reason why annotation metadata was not passed?

Reply:
Probably a mistake

Reply:
It was mainly used from tests, removed it https://github.com/micronaut-projects/micronaut-data/pull/2323/commits/df6d6db6ae543edd554965c5d51e48315a2ceaa5